src/main/java/io/confluent/connect/jdbc/sink/RecordValidator.java

```diff
@@ -0,0 +1,213 @@
+                + "with a %s value and %s value schema.",
+                config.connectorName(),
+                JdbcSinkConfig.DELETE_ENABLED,
+                config.deleteEnabled,
+                JdbcSinkConfig.PK_MODE,
+                config.pkMode.toString().toLowerCase(),
+                record.topic(),
+                record.kafkaPartition(),
+                record.kafkaOffset(),
+                record.timestamp(),
+                StringUtils.valueTypeOrNull(record.value()),
+                StringUtils.schemaTypeOrNull(record.valueSchema())
+            )
+        );
+      }
+      Schema valueSchema = record.valueSchema();
+      if (valueSchema == null) {
+        throw new ConnectException(
+            String.format(
+                "The JDBC sink connector '%s' is configured with '%s=%s' and '%s=%s' "
+                + "and therefore requires non-null record values that are structs, "
+                + "but found record at (topic='%s',partition=%d,offset=%d,timestamp=%d) "
+                + "with a %s value and %s value schema.",
+                config.connectorName(),
+                JdbcSinkConfig.DELETE_ENABLED,
+                config.deleteEnabled,
+                JdbcSinkConfig.PK_MODE,
+                config.pkMode.toString().toLowerCase(),
+                record.topic(),
+                record.kafkaPartition(),
+                record.kafkaOffset(),
+                record.timestamp(),
+                StringUtils.valueTypeOrNull(record.value()),
+                StringUtils.schemaTypeOrNull(record.valueSchema())
+            )
+        );
+      }
+      if (valueSchema.type() != Schema.Type.STRUCT) {
+        throw new ConnectException(
+            String.format(
+                "The JDBC sink connector '%s' is configured with '%s=%s' and '%s=%s' "
+                + "and therefore requires record values to be structs, "
+                + "but found record at (topic='%s',partition=%d,offset=%d,timestamp=%d) "
+                + "with a %s value and %s value schema.",
+                config.connectorName(),
+                JdbcSinkConfig.DELETE_ENABLED,
+                config.deleteEnabled,
+                JdbcSinkConfig.PK_MODE,
+                config.pkMode.toString().toLowerCase(),
+                record.topic(),
+                record.kafkaPartition(),
+                record.kafkaOffset(),
+                record.timestamp(),
+                StringUtils.valueTypeOrNull(record.value()),
+                StringUtils.schemaTypeOrNull(record.valueSchema())
+            )
+        );
+      }
+    };
+  }
+
+  static RecordValidator requiresKey(JdbcSinkConfig config) {
+    return record -> {
+      if (record.key() == null) {
+        throw new ConnectException(
+            String.format(
+                "The JDBC sink connector '%s' is configured with '%s=%s' and '%s=%s' "
+                + "and therefore requires non-null record keys, "
```

Comment:
WDYT about combining all of these throws together and parameterizing the message?
It looks like all of these functions are identical except for the `if` condition and this line in the error message.
```suggestion
                + "and %s, "
```

Reply:
I tried that, and it made the code more difficult to understand (since the entire error messages were never constructed in one place), and searching the code for an error message becomes more difficult. I think the clarity outweighs any duplicate trivial code.

Reply:
When I see so much similar code, I assume that there are more differences than there actually are, and play spot-the-difference. I don't need to do this if there is less code.

w.r.t. searching for error messages: The error message is already mostly un-searchable with so much string formatting. If this is valuable (i think it is), then maybe we can take another step to make this more searchable:

If the message had the parameterized message first, such as "Found record with null record keys: The JDBC sink connector is configured...." then I can search for the phrase before the colon, and find a static string in the class.